Snippet rework

Jarcho opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/14724)*

This is a further rework of our snippet accessing/creating code.

The general design goals with this:
* Avoid allocating when not strictly necessary.
* Make it easy to avoid linting if an invalid span is used.
* Make it easier to debug when an invalid span is used.
* Make it easy to minimize the number of lookups in the source map.
* Make it easy to avoid creating intermediary compressed spans.
* Make it as easy as possible to compose span adjustments/checks while maintaining the previous goals.

changelog: None

Jarcho commented:
The first two commits are mainly renames. The third commit has all the actual changes.

Jarcho commented:
Looking at `MISSING_DOCS_IN_PRIVATE_ITEMS`, that lint needs to be restructured a bit. Macros currently mess up it's implementation quite a bit.

Jarcho commented on clippy_utils/src/source.rs:
```diff
@@ -205,123 +257,274 @@ where
-    sm.ensure_source_file_source_present(&start.sf);
-    let range = start.pos.to_usize()..end.pos.to_usize();
-    Some(SourceFileRange { sf: start.sf, range })
+/// Handle to a source file's text and a range within that file.
+///
+/// With debug assertions the range is checked to be a valid substring of the source text. Without
+/// assertions `None` will be returned from various functions when accessing the substring of the
+/// source text fails.
+#[derive(Clone)]
+pub struct SourceFileRange {
+    file: SourceText,
+    range: Range<RelativeBytePos>,
 }
```
Doing so would require that `SourceFileRange ` always validate it's range. The main point of `SourceText` is that it's definitely a valid string.

They're also serving two different purpose. One is a substring of the source text, the other is a movable view of a whole file.

Alexendoo replied:
Always validating doesn't seem so bad to me, API wise it's some extra `?`ing when using `with_lo`/`with_hi`

Jarcho commented on clippy_lints/src/ranges.rs:
```diff
@@ -283,13 +283,14 @@ fn check_possible_range_contains(
-        && new_span.check_source_text(cx, |src| {
+        && new_lhs.span.ctxt() == ctxt
+        && let Some(span) = new_lhs.span.map_span(cx, |file| {
+            let src = file.with_hi(span.hi()).src_text()?;
             // Do not continue if we have mismatched number of parens, otherwise the suggestion is wrong
-            src.matches('(').count() == src.matches(')').count()
+            (src.matches('(').count() == src.matches(')').count()).then_some(file)
         })
```
Yeah, those need to be renamed.

Jarcho commented:
Should be ready, but still waiting on #14741 to be merged. Recent changes:
* Debug assert message now has a lot more info. The amount of info added is quite helpful when debugging span issues. `SourceMap` has to be threaded through when debug assertions are enabled. to make the cross file notes.
* `#[track_caller]` is used to get the panic location to give a useful position. This is way easier then reading the stack trace.
* Setting the start and end are split into `if_within` and `if_before/after` versions. The caller should always know which one should happen, but macros can cause tokens to get rearranged in fun ways.

I'm still very much against trying to merge `SourceText` and `SourceFileRange`. They model notably different things.

Jarcho committed:
Rework `clippy_utils::source`.


Rename `get_source_text` and `check_source_text` to `get_text` and `check_text`.

Jarcho commented:
The API has changed again; this time to support splitting a range. Changes include:
* `get_source_text` has been renamed to `get_text`. The "source" part of the name doesn't really add anything and just makes it longer. `check_source_text` was renamed similarly.
* `SourceFileRange` has been replaced with `SpanEditCx` which no longer contains the range.
* `map_range`'s callback now takes both the edit context and the range as separate parameters and returns either a single range or an array of ranges.
* Functions for shrinking a range based on it's contents have been replaced by a general `&str -> &str` transformation that recalculates the range positions. This can also return an array of strings.
* Some entry points no longer load external sources. `get_text` and `check_text` are still loading external sources since they hit debug assertions otherwise.
* A couple of general utils have been added (`display` and `StrExt`). They aren't specific to touching the source code, but they will mainly be used when working with it.

I might switch the range editing functions to be on the range, but that means an extra trait will need to be imported everywhere. `range.with_leading_prefix(scx, "prefix")` is nicer when chaining multiple, so it's probably worth it.

Alexendoo commented on clippy_utils/src/source.rs:
```diff
@@ -55,158 +93,563 @@ impl<'sm> HasSourceMap<'sm> for &LateContext<'sm> {
+    /// |`0`    |`m1!`, `m2!`|None    |
+    #[inline]
+    #[must_use]
+    fn walk_to_ctxt(self, ctxt: SyntaxContext) -> Option<Span>
```
This could do with some documentation on how it differs from `hygiene::walk_chain`

Alexendoo commented on clippy_utils/src/source.rs:
```diff
@@ -55,158 +93,563 @@ impl<'sm> HasSourceMap<'sm> for &LateContext<'sm> {
+            if expn.call_site.ctxt() != ctxt {
+                let sp = hygiene::walk_chain(expn.call_site, ctxt);
+                (sp.ctxt() == ctxt).then_some(sp)
+            } else if matches!(expn.kind, ExpnKind::Desugaring(DesugaringKind::RangeExpr)) {
```
What's unique about range desugaring that it's handled here? Should it be changed in rustc instead?

If I'm reading it right it also only applies when the outermost expn is said desugaring, what about a range in a macro?

Alexendoo commented on clippy_utils/src/source.rs:
```diff
@@ -216,81 +659,747 @@ where
+    where
+        P: Pattern,
+        for<'a> P::Searcher<'a>: ReverseSearcher<'a>;
+}
```
Some examples would be great

Alexendoo commented on clippy_utils/src/source.rs:
```diff
@@ -1,201 +1,602 @@
+}
+
+/// Helper functions to interact with the source text of a span/range.
+pub trait SpanExt: Sized {
```
Could do with a `Result`-style `See the module documentation for details.` link here

hkBst commented on clippy_lints/src/floating_point_arithmetic/custom_abs.rs:
```diff
@@ -14,11 +14,15 @@ use super::SUBOPTIMAL_FLOPS;
 /// is nonnegative.
 /// Used for check-custom-abs function below
-fn is_testing_positive(cx: &LateContext<'_>, expr: &Expr<'_>, test: &Expr<'_>) -> bool {
+fn is_testing_positive(cx: &LateContext<'_>, ctxt: SyntaxContext, expr: &Expr<'_>, test: &Expr<'_>) -> bool {
```
It seems confusing that there are now two different abbreviations of "context" for two different contexts. Maybe it would be clearer if `ctxt` became `syn_cx` for example.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/fee910077667df74b962aa66015398a9540f78d3..df7e99666ffa39c89b613fc48e35117f950dcc7a)*

ada4a replied:
Tbf `cx` is basically always the name we use for `Early/LateContext`, as is `ctxt` for `SyntaxContext`. So this shouldn't be confusing if you work with the codebase often.. Sorry, this sounds kind of gate-keepy, but I do believe it's true

hkBst replied:
@ada4a Thanks, that's totally fine! I would say that in that case it may still be something to address, but not something to hold against this change.

And I'm definitely looking at this as someone who hasn't worked extensively with the code, so I realize my view is biased in that way...

Jarcho replied:
This is used widely and with a very high level of consistency in rustc/clippy (clippy is worse for using `cx` and `ctxt` simultaneously). It definitely does have some WTFness to it, but changing it is probably more effort than it's worth with how widespread it is.

hkBst commented on clippy_utils/src/source.rs:
```diff
@@ -106,7 +106,7 @@ impl IntoSpan for Range<BytePos> {
 }
 
-pub trait SpanRangeExt: SpanRange {
+pub trait SpanExt: SpanRange {
```
I like short names, but I believe it is customary to add `Ext` to the trait one is *ext*ending. So unless we want to also rename `SpanRange` to `Span`, I'm feeling at least ambiguous about this change.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/fee910077667df74b962aa66015398a9540f78d3..df7e99666ffa39c89b613fc48e35117f950dcc7a)*

hkBst commented on clippy_lints/src/lib.rs:
```diff
@@ -1,15 +1,18 @@
+    rustc_private,
+    stmt_expr_attributes,
+    unwrap_infallible
+)]
```
This mixes a formatting change with a change in the listed elements.

I'm pointing this out as a clear example of what makes a proposed change harder than necessary to review.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/fee910077667df74b962aa66015398a9540f78d3..df7e99666ffa39c89b613fc48e35117f950dcc7a)*

hkBst commented:
If I may hazard a recommendation I would propose to separate the renaming commits, since they are relatively easy decisions but still contribute significantly to the +- figures. This would be a way to get those simple changes applied relatively quickly.

The last commit seems the be the one with lots of changes that are more unclear and would probably need more explanation of its merits. If possible I would definitely chop this up into bite-sized bits.

I would hope that this procedure would help in getting some of this reviewed/merged instead of sitting in the review queue, which is ultimately what everyone here wants.

Jarcho commented:
Ping @Alexendoo.

The special handling of range desugarings can hopefully be removed at point. A lot of our suggestion creation relies on getting the parenthesis surrounding an expression, but the call site of the desugaring does not have them.

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -1,25 +1,95 @@
+//! * [`SpanExt::check_text`]: A slightly simpler way to check a predicate on the text than using
+//!   `get_text`. Prefer this if possible.
+//! * [`SpanExt::map_range`]: The main way to adjust the range portion of a span or to split a
+//!   single span. See [`SpanEditCx`] for range adjustment utilities.
```
> or to split a single span

afaict it always returns one span only -- this part of the docs might be outdated?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

Jarcho replied:
Yes. That got split into a separate function.

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -1,25 +1,95 @@
+//! nodes in the AST or HIR representing a macro call (AST nodes have a variant for these, but it's
+//! only used before expansion). Because of this there's no direct way of knowing where a macro call
+//! occurred. Instead the `Span`, or more specifically the [`SyntaxContext`], contains which
+//! expansion created the node.
```
it seems like some words are missing here?
```suggestion
//! occurred. Instead the `Span`, or more specifically the [`SyntaxContext`], contains
//! information about which expansion created the node.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -1,25 +1,95 @@
+//! You _cannot_ assume anything about the `Span` or source text of any item. The parser will apply
+//! token substitution in some cases (e.g. replacing `（`, with `(`), macros can rearrange tokens,
+//! and proc-macros in particular can freely set the `Span` of any token to a different one. These
+//! can only be detected by checking the source text. With this the source text of all AST/HIR item
```
```suggestion
//! can only be detected by checking the source text. With this the source text of any AST/HIR item
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -55,158 +125,512 @@ impl<'sm> HasSourceMap<'sm> for &LateContext<'sm> {
-        } else {
-            None
-        }
+mod source_text {
+    use rustc_span::SourceFile;
+    use std::sync::Arc;
+
+    /// Handle to a substring of text in a source file.
+    #[derive(Clone)]
+    pub struct SourceText {
+        file: Arc<SourceFile>,
+        // This is a pointer into the text owned by the source file. If the source is external
+        // then the `FreezeLock` on the text must be frozen.
+        text: *const str,
     }
+    impl SourceText {
```
It wasn't immediately clear to me why we have this private `source_text` module with `SourceText` that gets immediately re-exported, and the two separate impl blocks, as the one inside the module has all its methods marked public anyway. Then I realized that this is probably done to restrict access to its private fields. Would you be okay with adding some comments explaining this?
```suggestion
/// We define [`SourceText`] inside this module to avoid exposing its private fields.
mod source_text {
    use rustc_span::SourceFile;
    use std::sync::Arc;

    /// Handle to a substring of text in a source file.
    #[derive(Clone)]
    pub struct SourceText {
        file: Arc<SourceFile>,
        // This is a pointer into the text owned by the source file. If the source is external
        // then the `FreezeLock` on the text must be frozen.
        text: *const str,
    }

    /// This impl block needs to be inside the module in order to be able to access the private
    /// fields.
    impl SourceText {
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

Jarcho replied:
The main uses for using an inline module like this are unnameable public traits, data structures with unsafe code and cfg attributes. You can tell which this case is from the raw pointer field and/or unsafe code in the methods. I won't blame you for not having seen a case like this if you don't interact with unsafe code, but this isn't an esoteric use of a module. Commenting a single instance of this isn't even a big deal, but the logic that leads to doing so leads to doing it every time it comes up (and for different patterns somebody might not recognize).

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -216,81 +640,768 @@ where
+            parent: data.parent,
+        };
+        scx.dbg_check_range(None, RelativeBytePos(sfp.pos.0)..end);
+        Some((scx, RelativeBytePos(sfp.pos.0)..end))
```
nit: you already create a variable for `end`, but the start part is identical between the two expressions as well -- maybe just create a variable for the whole range at this point?
```suggestion
        let range = RelativeBytePos(sfp.pos.0)..RelativeBytePos(data.hi.0.wrapping_sub(sfp.sf.start_pos.0));

        let scx = Self {
            text: SourceText::for_local_file(sfp.sf)?,
            ctxt: data.ctxt,
            parent: data.parent,
        };
        scx.dbg_check_range(None, range);
        Some((scx, range))
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -216,81 +640,768 @@ where
+    #[must_use]
+    pub fn contains_pos(&self, pos: BytePos) -> bool {
+        let file = self.file();
+        pos.0.wrapping_sub(file.start_pos.0) <= file.normalized_source_len.0
```
nit: I _think_ you can use [`SourceFile::contains`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.SourceFile.html#method.contains) here?
```suggestion
        self.file().contains(pos)
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

ada4a commented on clippy_lints/src/implicit_hasher.rs:
```diff
@@ -118,16 +118,19 @@ impl<'tcx> LateLintPass<'tcx> for ImplicitHasher {
-                        if let Some(range) = range {
-                            range.with_ctxt(item.span.ctxt())
+                    let generics_suggestion_span = if impl_.generics.span.is_dummy() {
+                        if let Some(sp) = item.span.map_range(cx, |scx, range| {
+                            range
+                                .shrink_end_to(scx, target.span().lo_ctxt())?
+                                .map_range_text(scx, |src| src.split_once("impl").map(|(_, x)| x))
+                        }) {
+                            sp
                         } else {
                             return;
                         }
-                    });
+                    } else {
+                        impl_.generics.span
+                    };
```
is there a reason to not continue using `Span::substitute_dummy` here?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

Jarcho replied:
That requires evaluating the replacement before knowing we have a dummy.

ada4a replied:
That was also the case before this PR though, no? Miscellaneous optimisations like this one are welcome of course, but putting them into a separate commit would help review..

PS: maybe it'd make sense to have a lazy version of this function, something like `Span::substitute_dummy_wuth`

Jarcho replied:
It's actually not an optimization here. The span creation fails under more conditions now and has to be delayed.

ada4a commented on clippy_lints/src/implicit_hasher.rs:
```diff
@@ -164,19 +167,24 @@ impl<'tcx> LateLintPass<'tcx> for ImplicitHasher {
-                                range.with_ctxt(item.span.ctxt())
+
+                        let generics_suggestion_span = if generics.span.is_dummy() {
+                            if let Some(sp) = item.span.map_range(cx, |scx, range| {
+                                range
+                                    .shrink_end_to(scx, body.params[0].pat.span.lo_ctxt())?
+                                    .map_range_text(scx, |src| {
+                                        src.split_once("fn")
+                                            .and_then(|(_, x)| x.split_once('('))
+                                            .map(|(_, x)| x)
+                                    })
+                            }) {
+                                sp
                             } else {
                                 return;
                             }
-                        });
+                        } else {
+                            generics.span
+                        };
```
ditto

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

ada4a commented on clippy_utils/src/source.rs:
````diff
@@ -216,81 +640,768 @@ where
-        && let Some(text) = &src.sf.src
-        && let Some(range) = f(&src.sf, text, src.range.clone())
+pub trait StrExt {
+    /// Gets the substring which ranges from the start of the first match of the pattern to the end
+    /// of the second match. Returns `None` if the pattern doesn't occur twice.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # #![feature(rustc_private)]
+    /// # use clippy_utils::source::StrExt;
+    /// let s = "move |arg| arg.foo()";
+    /// assert_eq!(s.find_bounded_inclusive('|'), Some("|arg|"));
+    ///
+    /// let s = "foo | bar";
+    /// assert_eq!(s.find_bounded_inclusive('|'), None)
+    /// ```
+    fn find_bounded_inclusive(&self, pat: impl Pattern) -> Option<&Self>;
````
To be completely honest I'm not sure this function pulls its weight, at least as it is now: conceptually, it extracts the portion of the text between some delimiters, but pretty much always, the start and end symbols of a delimiter are actually distinct, e.g. `(` and `)` -- `|` is more of an exception. So maybe this function should take two patterns instead?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

Jarcho replied:
I don't particularly like this function. It can't be used for anything which may nest which would account for basically everything you might want to use this for. Even where it's used it may not work since `|` can appear inside a closures arguments as a type constant expression.

I can try to remove the need for this separately, but I'm not really inclined to make this nice. It already has more written about it than it deserves.

ada4a commented on clippy_utils/src/source.rs:
```diff
@@ -55,158 +125,512 @@ impl<'sm> HasSourceMap<'sm> for &LateContext<'sm> {
+        #[inline]
+        #[must_use]
+        pub fn as_str(&self) -> &str {
+            // SAFETY: `text` is owned by `sf` and comes from either an `Option<Arc<String>>`, or a
```
What's `sf`? Do you mean `file`?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14724/d0a919d7c850dd66f2be66ae7a6f09fb099f38be..c92dc5ccf5db7a5f3a3d2a44e7d096aa441ca00c)*

Jarcho replied:
Yes. That name is outdated.
